### PR TITLE
[Issue #4572]-Allow users to leave an organization

### PR DESCRIPTION
**Pull request by nayan-rafiq**
## Description
Currently, an user can not remove himself from an organization. The admins can remove other members. After this change, admin users will be able to remove themselves also if there is minimum one admin. Also other users will have a leave organization option from where they can leave the organization.

Fixes # (issue)
Issue #4572 

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Unit tests
- Tested on local machine

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/allow-admin-leave-organization 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 51.57 **(-0.01)** | 32.59 **(0)** | 29.75 **(-0.01)** | 52.09 **(0)**
 :red_circle: | app/client/src/actions/userActions.ts | 68.18 **(-0.11)** | 100 **(0)** | 0 **(0)** | 84.21 **(-4.03)**
 :red_circle: | app/client/src/api/UserApi.tsx | 64.58 **(0.14)** | 0 **(0)** | 6.67 **(-0.47)** | 63.83 **(0.19)**
 :green_circle: | app/client/src/sagas/userSagas.tsx | 17.71 **(0.58)** | 1.92 **(-0.08)** | 6.25 **(-0.42)** | 21.12 **(0.73)**</details>

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java [hunk 1]
```diff
@@ -90,6 +90,12 @@ public UserController(UserService service,
                 .map(user -> new ResponseDTO<>(HttpStatus.OK.value(), user, null));
     }
 
+    @DeleteMapping("/organization/{orgId}")
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/UserController.java:93]: We are currently using the `@DeleteMapping` only for delete/archive operations right? So this feels a little inconsistent. This feels like a _change_ to the organization, and the organization itself doesn't disappear for everyone. I feel a `PUT` request with an endpoint like `/leaveOrganization/{orgId}` is more consistent, and is very similar to the method just above, which defines `PUT /addOrganization/{orgId}`.

File: app/client/src/pages/Applications/index.tsx [hunk 2]
```diff
@@ -623,58 +628,66 @@ function ApplicationsSection(props: any) {
...
+                          OrgNameChange(value, organization.id);
+                        }}
+                        placeholder="Workspace name"
+                        savingState={
+                          isSavingOrgInfo
+                            ? SavingState.STARTED
+                            : SavingState.NOT_STARTED
+                        }
+                        underline
+                      />
```
Comment (pskanade) [inline app/client/src/pages/Applications/index.tsx:660]: For consistency `OrgName` should be shown to all users. And it should be editable only for user with `hasManageOrgPermissions: true`

**Review by pskanade — CHANGES_REQUESTED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/UserOrganizationServiceImpl.java [hunk 3]
```diff
@@ -254,6 +262,83 @@ public UserOrganizationServiceImpl(SessionUserService sessionUserService,
...
+                        return emailMono
+                                .thenReturn(addedOrganization);
+                    });
+                } else {
+                    // If the roleName was not present, then it implies that the user is being removed from the org.
+                    // Since at this point we have already removed the user from the organization,
+                    // remove the organization from recent org list of UserData
+                    // we also need to remove the org id from User.orgIdList
+                    finalUpdatedOrganizationMono = userRemovedOrganizationMono.flatMap(userDataRepository
+                            .removeOrgFromRecentlyUsedList(user.getId(), organization.getId())::thenReturn)
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/UserOrganizationServiceImpl.java:316]: Minor refactoring to read this better.

```suggestion
                    finalUpdatedOrganizationMono = userDataRepository
                            .removeOrgFromRecentlyUsedList(user.getId(), organization.getId())
                            .then(userRemovedOrganizationMono)
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/UserOrganizationServiceImpl.java [hunk 4]
```diff
@@ -254,6 +262,83 @@ public UserOrganizationServiceImpl(SessionUserService sessionUserService,
...
+                            .flatMap(organization1 -> {
+                                    if(user.getOrganizationIds() != null) {
+                                        user.getOrganizationIds().remove(organization.getId());
+                                        return userRepository.save(user).thenReturn(organization1);
+                                    }
+                                    return Mono.just(user).thenReturn(organization1);
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/UserOrganizationServiceImpl.java:322]:
```suggestion
                                    return Mono.just(organization1);
```

**Comment by sharat87** [comment]
@nayan-rafiq, looks good. Just a few minor things and then we should be able to merge this in. Thanks!

File: app/client/src/pages/Applications/index.tsx (hunk 2 above)
Comment (nayan-rafiq) [reply app/client/src/pages/Applications/index.tsx:660]: The user has just clicked on the org name and a popup is shown. Showing the orgname again as first option in the popup without the edit button will be confusing IMO.

File: app/client/src/constants/ReduxActionConstants.tsx [hunk 5]
```diff
@@ -366,6 +366,8 @@ export const ReduxActionTypes: { [key: string]: string } = {
   FETCH_ALL_ROLES_INIT: "FETCH_ALL_ROLES_INIT",
   DELETE_ORG_USER_INIT: "DELETE_ORG_USER_INIT",
   DELETE_ORG_USER_SUCCESS: "DELETE_ORG_USER_SUCCESS",
+  DELETE_ORG_USER_ERROR: "DELETE_ORG_USER_ERROR",
```
[the remark this answers is not in the document]
Comment (nayan-rafiq) [reply app/client/src/constants/ReduxActionConstants.tsx:369]: Got it.

File: app/client/src/pages/Applications/index.tsx [hunk 6]
```diff
@@ -623,58 +628,66 @@ function ApplicationsSection(props: any) {
...
-                      })
-                    }
-                    text="Organization Settings"
-                  />
-                  <MenuItem
-                    icon="share"
-                    onSelect={() => setSelectedOrgId(organization.id)}
-                    text="Share"
-                  />
+                  {hasManageOrgPermissions && (
```
[the remark this answers is not in the document]
Comment (nayan-rafiq) [reply app/client/src/pages/Applications/index.tsx:638]: Added cypress test for the feature

**Comment by nayan-rafiq** [comment]
> @nayan-rafiq, looks good. Just a few minor things and then we should be able to merge this in. Thanks!

@sharat87 I've applied the requested changes. Would you please have a look again?

**Review by pranavkanade — APPROVED** [review]
FE LGTM! Approving for tests

**Review by sharat87 — APPROVED** [review]

**Pull request merged**